src/main/kotlin/org/rust/lang/core/macros/RsExpandedElement.kt

```diff
@@ -58,13 +56,31 @@ val RsExpandedElement.expandedFromRecursively: RsMacroCall?
     }
 
 fun PsiElement.findMacroCallExpandedFrom(): RsMacroCall? {
-    val found = ancestors
+    val found = findMacroCallExpandedFromNonRecursive()
+    return found?.findMacroCallExpandedFrom() ?: found
+}
+
+fun PsiElement.findMacroCallExpandedFromNonRecursive(): RsMacroCall? {
+    return ancestors
         .filterIsInstance<RsExpandedElement>()
-        .mapNotNull { it.expandedFromRecursively }
+        .mapNotNull { it.expandedFrom }
         .firstOrNull()
-    return found?.findMacroCallExpandedFrom() ?: found
 }
 
+val PsiElement.isExpandedFromMacro: Boolean
+    get() = findMacroCallExpandedFromNonRecursive() != null
+
+/**
+ * If receiver element is inside a macro expansion, returns the element inside the macro call
+ * we should navigate to (or the macro call itself if there isn't such element inside a macro call).
+ * Returns null if the element isn't inside a macro expansion
+ */
+fun PsiElement.findNavigationTargetIfMacroExpansion(): PsiElement? =
+    /** @see RsNamedElementImpl.getTextOffset */
+    ((this as? RsNameIdentifierOwner)?.nameIdentifier ?: this).findElementExpandedFrom()
+        // fallback to macro call
+        ?: this.findMacroCallExpandedFrom()?.path
```

Comment:
```suggestion
        ?: findMacroCallExpandedFrom()?.path
```

Reply:
Maybe rewrite it somehow? For example, in the following way 
```kotlin
fun PsiElement.findNavigationTargetIfMacroExpansion(): PsiElement? {
    /** @see RsNamedElementImpl.getTextOffset */
    val element = (this as? RsNameIdentifierOwner)?.nameIdentifier ?: this
    return element.findElementExpandedFrom() ?: findMacroCallExpandedFrom()?.path // fallback to macro call
}
```
It seems it's more readable